lib/HLSL/HLOperationLower.cpp

```diff
@@ -9493,16 +9491,24 @@ void LowerRecordAccessToGetNodeRecordPtr(HLModule &HLM) {
     if (F->user_empty())
       continue;
     hlsl::HLOpcodeGroup group = hlsl::GetHLOpcodeGroup(F);
-    if (group == HLOpcodeGroup::HLIntrinsic) {
+    if (group == HLOpcodeGroup::HLSubscript) {
       for (auto U = F->user_begin(); U != F->user_end();) {
         Value *User = *(U++);
         if (!isa<Instruction>(User))
           continue;
         // must be call inst
         CallInst *CI = cast<CallInst>(User);
-        IntrinsicOp opcode = static_cast<IntrinsicOp>(hlsl::GetHLOpcode(CI));
-        if (opcode != IntrinsicOp::IOP_ExtractRecordStructFromArray)
+        HLSubscriptOpcode opcode =
+            static_cast<HLSubscriptOpcode>(hlsl::GetHLOpcode(CI));
+        if (opcode != HLSubscriptOpcode::DefaultSubscript)
+          continue;
+
+        hlsl::OP *OP = &helper.hlslOP;
+        Value *Handle = CI->getArgOperand(HLOperandIndex::kHandleOpIdx);
+        if (Handle->getType() != OP->GetNodeRecordHandleType()) {
```

Comment:
Should this ever happen? Maybe there should be an assert in there as well?

Reply:
This condition can be true on non-NodeRecordHandle types.  DxilGenerationPass.cpp:runOnModule will unconditionally call LowerRecordAccessToGetNodeRecordPtr without checking the record type. One could argue that there should be logic preventing that function from being called in the caller function when the handle type is not a node record, but that is a separate issue. So, I think there should not be any assert here.